packages/node/src/behavior/context/server/OnlineContext.ts

```diff
@@ -4,21 +4,33 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-import { AccessControlServer } from "#behaviors/access-control";
 import { Agent } from "#endpoint/Agent.js";
 import { Endpoint } from "#endpoint/Endpoint.js";
 import { EndpointType } from "#endpoint/type/EndpointType.js";
-import { Diagnostic, ImplementationError, InternalError, MaybePromise, Transaction } from "#general";
+import { AsyncObservable, Diagnostic, ImplementationError, InternalError, MaybePromise, Transaction } from "#general";
 import { AccessLevel } from "#model";
 import type { Node } from "#node/Node.js";
 import type { Message, NodeProtocol } from "#protocol";
-import { AccessControl, AclEndpointContext, MessageExchange, SecureSession, Subject } from "#protocol";
+import {
+    AccessControl,
+    AclEndpointContext,
+    FabricAccessControlManager,
+    MessageExchange,
+    SecureSession,
+    Subject,
+} from "#protocol";
 import { FabricIndex, NodeId } from "#types";
 import { ActionContext } from "../ActionContext.js";
 import { Contextual } from "../Contextual.js";
 import { NodeActivity } from "../NodeActivity.js";
 import { ContextAgents } from "./ContextAgents.js";
 
+/**
+ * Caches completion events per exchange. Uses if multiple OnlineContext instances are created for an exchange.
+ * Entries will be cleaned up when the exchange is closed.
+ */
+const exchangeCompleteEvents = new Map<MessageExchange, AsyncObservable<[session?: ActionContext | undefined]>>();
```

Comment:
Not sure it's necessary but could consider using a weak map here so we don't have a leak if an exchange fails to report as complete